File: forza/domain/normalizer.py
from __future__ import annotations

import re
import unicodedata
import difflib
import logging
from pathlib import Path
from dataclasses import dataclass

from .text_utils import load_nonempty_lines


log = logging.getLogger("forza")


# ── Reference data container ──────────────────────────────────────────────────

@dataclass
class ReferenceData:
    tracks:  list[str]
    cars:    list[str]
    car_map: dict[str, str]   # normalised key → original name


def load_reference_seed_text_data(tracks_file: Path, cars_file: Path) -> ReferenceData:
    """Load explicit seed/test reference text files into a ReferenceData value."""
    tracks = load_nonempty_lines(tracks_file, warn_missing=True, logger=log)
    cars = load_nonempty_lines(cars_file, warn_missing=True, logger=log)
    return ReferenceData(tracks=tracks, cars=cars, car_map=_build_car_map(cars))


def _build_car_map(cars: list[str]) -> dict[str, str]:
    result: dict[str, str] = {}
    for car in cars:
        key = _normalise(car, spaces=False)
        if key not in result:
            result[key] = car
    return result


# ── Normalisation helpers ─────────────────────────────────────────────────────

def _normalise(text: str, spaces: bool = True) -> str:
    nfkd = unicodedata.normalize("NFKD", str(text))
    s = "".join(c for c in nfkd if not unicodedata.combining(c))
    if not spaces:
        s = re.sub(r"\W+", "", s)
    return s.lower().strip()


def _track_key(text: str) -> str:
    return re.sub(r"[^a-z0-9]+", " ", _normalise(text)).strip()


# ── Track correction ──────────────────────────────────────────────────────────

def fix_track_name(raw: str, refs: ReferenceData) -> str | None:
    """
    Match a raw OCR track name against the supplied reference tracks.

    Strategy (first match wins):
      1. Exact match (case-insensitive)
      2. Accent-normalised exact match
      3. Punctuation-insensitive exact match
      4. Unique prefix match — only when exactly one track starts with the
         prefix.  When multiple tracks share the same prefix (e.g. Le Mans
         Full Circuit vs Old Mulsanne Circuit both start with the same base
         name), returns None so the caller can flag for
         review instead of guessing the wrong layout.
      5. Fuzzy match (cutoff 0.75) — only when prefix match found zero
         candidates, to avoid masking ambiguous prefixes.
      6. Return original string unchanged (unrecognised, not ambiguous).

    Returns None when the result is ambiguous and choosing silently would risk
    returning the wrong layout. Empty raw input still returns the original empty
    string, so callers can distinguish missing text from ambiguous text.
    """
    if not raw or not refs.tracks:
        return raw

    term     = " ".join(raw.split()).strip()
    term_low = term.lower()

    # 1) Exact
    for track in refs.tracks:
        if track.lower() == term_low:
            return track

    # 2) Accent-normalised exact
    term_norm = _normalise(term)
    for track in refs.tracks:
        if _normalise(track) == term_norm:
            return track

    # 3) Punctuation-insensitive exact
    term_key = _track_key(term)
    for track in refs.tracks:
        if _track_key(track) == term_key:
            return track

    # 4) Prefix match — safe only when unambiguous
    prefix_matches = [t for t in refs.tracks if _track_key(t).startswith(term_key)]
    if len(prefix_matches) == 1:
        return prefix_matches[0]
    if len(prefix_matches) > 1:
        # Multiple layouts share this prefix — do NOT guess.
        log.debug(
            f"[normalizer] Ambiguous prefix '{term}' matches "
            f"{[t for t in prefix_matches]} — flagging for review"
        )
        return None

    # 5) Fuzzy — only when prefix found nothing (avoids masking ambiguous cases)
    matches = difflib.get_close_matches(term, refs.tracks, n=1, cutoff=0.75)
    if matches:
        return matches[0]

    # 6) Unrecognised — return unchanged so caller can decide
    return term


# ── Car correction ────────────────────────────────────────────────────────────

def fix_car_name(raw: str, refs: ReferenceData) -> str:
    """
    Match a raw OCR car name against the pre-computed car map.

    Strategy (first match wins):
      1. Normalised exact match  (O(1))
      2. Substring match
      3. Fuzzy match  (cutoff 0.85)
      4. Return original string unchanged
    """
    if not raw or not refs.car_map:
        return raw

    raw_str  = str(raw).strip()
    raw_norm = _normalise(raw_str, spaces=False)

    # 1) Exact
    if raw_norm in refs.car_map:
        corrected = refs.car_map[raw_norm]
        if corrected != raw_str:
            log.debug(f"[car] '{raw_str}' → '{corrected}'")
        return corrected

    # 2) Substring
    candidates = [v for k, v in refs.car_map.items() if raw_norm in k]
    if len(candidates) == 1:
        log.debug(f"[car] '{raw_str}' → '{candidates[0]}' (substring)")
        return candidates[0]

    # 3) Fuzzy
    matches = difflib.get_close_matches(
        raw_str, list(refs.car_map.values()), n=1, cutoff=0.85
    )
    if matches:
        log.debug(f"[car] '{raw_str}' → '{matches[0]}' (fuzzy)")
        return matches[0]

    log.debug(f"[car] no match for '{raw_str}' — keeping original")
    return raw_str

File: forza-rust/crates/forza-domain/src/normalizer.rs
//! Reference data container plus track/car name correction strategies.

use std::collections::{HashMap, HashSet};

use crate::difflib;
use crate::text_utils;

/// Fuzzy cutoff for track names (Python parity, see `difflib.get_close_matches`).
pub const TRACK_FUZZY_CUTOFF: f64 = 0.75;
/// Fuzzy cutoff for car names (stricter: car catalog is larger).
pub const CAR_FUZZY_CUTOFF: f64 = 0.85;

/// Pre-computed normalized forms of one canonical track name, so
/// [`fix_track_name`] never renormalizes the whole catalog per call
/// (previously O(rows · refs · nfkd_cost)).
#[derive(Debug, Clone)]
struct PreparedTrack {
    lower: String,
    norm: String,
    key: String,
    original: String,
}

/// Reference catalog: canonical track names, car names, and a normalized
/// car key map (normalized key → original name), first occurrence wins.
#[derive(Debug, Clone, Default)]
pub struct ReferenceData {
    pub tracks: Vec<String>,
    pub cars: Vec<String>,
    pub car_map: HashMap<String, String>,
    prepared_tracks: Vec<PreparedTrack>,
    car_values: Vec<String>,
}

impl ReferenceData {
    pub fn from_lines(tracks: Vec<String>, cars: Vec<String>) -> Self {
        let prepared_tracks = tracks
            .iter()
            .map(|t| PreparedTrack {
                lower: t.to_lowercase(),
                norm: normalize(t, true),
                key: track_key(t),
                original: t.clone(),
            })
            .collect();
        let (car_map, car_values) = build_car_map(&cars);
        Self {
            tracks,
            cars,
            car_map,
            prepared_tracks,
            car_values,
        }
    }
}

fn normalize(text: &str, spaces: bool) -> String {
    text_utils::normalize_ascii_compare(text, spaces)
}

fn track_key(text: &str) -> String {
    let norm = normalize(text, true);
    let mut out = String::with_capacity(norm.len());
    let mut pending_sep = false;
    for ch in norm.chars() {
        if ch.is_ascii_lowercase() || ch.is_ascii_digit() {
            if pending_sep {
                out.push(' ');
                pending_sep = false;
            }
            out.push(ch);
        } else {
            pending_sep = !out.is_empty();
        }
    }
    out
}

fn build_car_map(cars: &[String]) -> (HashMap<String, String>, Vec<String>) {
    let mut map: HashMap<String, String> = HashMap::with_capacity(cars.len());
    let mut seen: HashSet<String> = HashSet::with_capacity(cars.len());
    let mut values: Vec<String> = Vec::with_capacity(cars.len());
    for car in cars {
        let key = normalize(car, false);
        if seen.insert(key.clone()) {
            map.insert(key, car.clone());
            values.push(car.clone());
        }
    }
    (map, values)
}

/// Match a raw OCR track name against the reference list.
///
/// Returns `None` only when a prefix match is ambiguous; unrecognised input
/// is returned unchanged so callers can flag it.
pub fn fix_track_name(raw: &str, refs: &ReferenceData) -> Option<String> {
    if raw.is_empty() || refs.tracks.is_empty() {
        return Some(raw.to_string());
    }

    let term: String = raw.split_whitespace().collect::<Vec<_>>().join(" ");
    let term_low = term.to_lowercase();

    // 1) Exact
    for prepared in &refs.prepared_tracks {
        if prepared.lower == term_low {
            return Some(prepared.original.clone());
        }
    }

    // 2) Accent-normalised exact
    let term_norm = normalize(&term, true);
    for prepared in &refs.prepared_tracks {
        if prepared.norm == term_norm {
            return Some(prepared.original.clone());
        }
    }

    // 3) Punctuation-insensitive exact
    let term_key = track_key(&term);
    for prepared in &refs.prepared_tracks {
        if prepared.key == term_key {
            return Some(prepared.original.clone());
        }
    }

    // 4) Prefix match — safe only when unambiguous
    let prefix_matches: Vec<&String> = refs
        .prepared_tracks
        .iter()
        .filter(|t| t.key.starts_with(term_key.as_str()))
        .map(|t| &t.original)
        .collect();
    match prefix_matches.len() {
        1 => return Some(prefix_matches[0].clone()),
        n if n > 1 => return None,
        _ => {}
    }

    // 5) Fuzzy — only when prefix found nothing
    let matches = difflib::get_close_matches(&term, &refs.tracks, 1, TRACK_FUZZY_CUTOFF);
    if let Some(first) = matches.into_iter().next() {
        return Some(first);
    }

    // 6) Unrecognised — unchanged
    Some(term)
}

/// Match a raw OCR car name against the pre-computed car map.
pub fn fix_car_name(raw: &str, refs: &ReferenceData) -> String {
    if raw.is_empty() || refs.car_map.is_empty() {
        return raw.to_string();
    }

    let raw_str = raw.trim();
    let raw_norm = normalize(raw_str, false);
    // A punctuation-only input normalizes to "": `contains("")` would match
    // every key, so bail out early instead of relying on later stages.
    if raw_norm.is_empty() {
        return raw_str.to_string();
    }

    // 1) Exact (O(1) hash lookup, first occurrence wins at build time)
    if let Some(corrected) = refs.car_map.get(&raw_norm) {
        return corrected.clone();
    }

    // 2) Substring (unique only)
    let candidates: Vec<&String> = refs
        .car_map
        .iter()
        .filter(|(k, _)| k.contains(raw_norm.as_str()))
        .map(|(_, v)| v)
        .collect();
    if candidates.len() == 1 {
        return candidates[0].clone();
    }

    // 3) Fuzzy over values in insertion order (no per-call clone)
    if let Some(first) = difflib::get_close_matches(raw_str, &refs.car_values, 1, CAR_FUZZY_CUTOFF)
        .into_iter()
        .next()
    {
        return first;
    }

    raw_str.to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn refs() -> ReferenceData {
        ReferenceData::from_lines(
            vec![
                "Fuji Speedway GT".to_string(),
                "Fuji Speedway".to_string(),
                "Nürburgring Nordschleife".to_string(),
                "Le Mans Full Circuit".to_string(),
                "Le Mans Old Mulsanne Circuit".to_string(),
            ],
            vec!["Audi R8 LMS".to_string(), "BMW M4 GT3".to_string()],
        )
    }

    #[test]
    fn exact_and_normalized_track_match() {
        let r = refs();
        assert_eq!(
            fix_track_name("fuji speedway", &r).unwrap(),
            "Fuji Speedway"
        );
        assert_eq!(
            fix_track_name("Nurburgring Nordschleife", &r).unwrap(),
            "Nürburgring Nordschleife"
        );
    }

    #[test]
    fn ambiguous_prefix_returns_none_for_review() {
        let r = refs();
        assert_eq!(fix_track_name("Le Mans", &r), None);
        assert_eq!(
            fix_track_name("Le Mans Full", &r).unwrap(),
            "Le Mans Full Circuit"
        );
    }

    #[test]
    fn punctuation_insensitive_and_unrecognised() {
        let r = refs();
        assert_eq!(
            fix_track_name("nürburgring  nordschleife!", &r).unwrap(),
            "Nürburgring Nordschleife"
        );
        assert_eq!(
            fix_track_name("Totally Unknown Track", &r).unwrap(),
            "Totally Unknown Track"
        );
    }

    #[test]
    fn car_exact_normalised_match_is_case_insensitive_no_spaces() {
        let r = refs();
        assert_eq!(fix_car_name("audi r8 lms", &r), "Audi R8 LMS");
        assert_eq!(fix_car_name("audir8lms", &r), "Audi R8 LMS");
        assert_eq!(fix_car_name("Unknown Car", &r), "Unknown Car");
    }

    #[test]
    fn punctuation_only_input_never_matches_single_car_catalog() {
        let r = ReferenceData::from_lines(vec![], vec!["Audi R8 LMS".to_string()]);
        assert_eq!(fix_car_name("---", &r), "---");
        assert_eq!(fix_car_name("...", &r), "...");
    }
}
